doc/src/guides/contributing/deprecations.md

```diff
@@ -0,0 +1,388 @@
+indicating what the user should be using instead.
+
+This allows users a chance to update their code without it completely
+breaking. It also gives SymPy an opportunity to give users an informative
+message on how to update their code, rather than making their code simply
+error or start giving wrong answers.
+
+## When to deprecate and when not to
+
+Sometimes the API of SymPy must change in an incompatible way. We try to avoid
+this, because changing API breaks people's code, but it is often deemed worthy
+to do so. Some reasons APIs are changed can include:
+
+- The existing API is confusing.
+- There is unnecessary redundancy in the API.
+- The existing API limits what is possible.
+
+Because one of the core use-cases of SymPy is to be usable as a library, we
+take API breakage very seriously. Whenever an API breakage is necessary, the
+following steps should be taken:
+
+- Discuss the API change with the community. Be sure that the improved API is
+  indeed better, and worth a breakage. It is important to get API right so
+  that things will not need to break again to "fix" it a second time.
+- Document any API breakages in the relevant section of the release notes for
+  the relevant release. This section is typically near the top of the release
+  notes.
+- If it is possible, deprecate the old API.
+
+When considering whether a change requires a deprecation, two things must be
+considered:
+
+- Is the change backwards incompatible?
+- Is the behavior being changed public API?
+
+A change is backwards incompatible if user code making use of it would stop
+working after the change.
+
+What counts as "public API" needs to be considered on a case-by-case basis.
+The exact rules for what does and doesn't constitute public API for SymPy
+are still not fully codified. Very broadly, some thing that constitute public
+API are (this list is not exhaustive)
+
+- Function names.
+- Keyword argument names.
+- Keyword argument default values.
+- The mathematical conventions used to define a function.
+- Submodule names for **public** functions that are not included in the
+  top-level `sympy/__init__.py`, since users must use the fully qualified
+  module name to import them (note: many such functions are not included
+  because they aren't actually public API at all. For these, no deprecations
+  are required to make breaking changes).
+
+And broadly, a non-exhaustive list of some things that don't consistent public
+API, and therefore don't require deprecations to change, include
+
+- The precise form of an expression. In general, functions may be changed to
+  return a different but mathematically equivalent form of the same
+  expression. This includes a function returning a value which it was not able
+  to compute previously.
+- The name of a submodule for functions is included in the top-level
+  `sympy/__init__.py`. Such functions should be imported directly, like `from
+  sympy import <name>`.
```

Comment:
I'll make it clearer. It means something like the name of the submodule `sympy.functions.elementary.exponential` isn't public API if we wanted to move the files around, because all the public names in that should be imported from the top-level.